Added InstanceReferenceKind to IInstanceReferenceOperation for tellin…

333fred opened the pull request:
…g between this. usage and initializer usage.

Split out part of https://github.com/dotnet/roslyn/pull/25322 to reduce the resulting size of the PR, and update the base to current features/dataflow. I've addressed initial relevant comments from that review.

AlekseyTs commented:
@333fred It looks like there are legitimate test failures.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/InstanceReferenceKind.cs:
```diff
@@ -0,0 +1,20 @@
+    public enum InstanceReferenceKind
+    {
+        /// <summary>
+        /// Reference to the containing type. Used for <code>this</code> and <code>base</code> in C# code, and <code>This</code>,
```
>This [](start = 117, length = 4)

Did you mean to say "Me" here?

AlekseyTs replied:
>Reference to the containing type. [](start = 12, length = 33)

"Reference to **an instance of** the containing type." ?

AlekseyTs commented on src/Compilers/Core/Portable/Operations/InstanceReferenceKind.cs:
```diff
@@ -0,0 +1,20 @@
+        /// </summary>
+        ContainingTypeInstance,
+        /// <summary>
+        /// Reference to a member of the object being initialized in a C# object or collection initializer, or a VB With statement.
```
>Reference to a member [](start = 12, length = 21)

This doesn't sound right. I don't think IInstanceReferenceOperation ever references a member.

CyrusNajmabadi commented:
> "and initializer usage."

Can you give an example of what you mean by that?  Thanks!

AlekseyTs commented:
>     /// Current usage:

It looks like this list here should be expanded with more items

---
Refers to: src/Compilers/Core/Portable/Operations/IInstanceReferenceExpression.cs:8 in 33db5e0. [](commit_id = 33db5e098fbe4821b745356988ee3e00c16a1709, deletion_comment = False)

AlekseyTs commented on src/Compilers/Core/Portable/PublicAPI.Unshipped.txt:
```diff
@@ -10,30 +10,15 @@ Microsoft.CodeAnalysis.MetadataImportOptions
 Microsoft.CodeAnalysis.MetadataImportOptions.All = 2 -> Microsoft.CodeAnalysis.MetadataImportOptions
 Microsoft.CodeAnalysis.MetadataImportOptions.Internal = 1 -> Microsoft.CodeAnalysis.MetadataImportOptions
 Microsoft.CodeAnalysis.MetadataImportOptions.Public = 0 -> Microsoft.CodeAnalysis.MetadataImportOptions
+Microsoft.CodeAnalysis.OperationKind.CaughtException = 94 -> Microsoft.CodeAnalysis.OperationKind
```
In this file, I would expect to see only changes around IInstanceReferenceOperation.ReferenceKind property and InstanceReferenceKind type. Please revert all other unnecessary changes.

333fred replied:
The codefix that adds things to the public api automatically sorts this file. If we don't do this now, we'll have to manually fix this file up every time we want to add anything new. I'm going to leave this as is.

---
In reply to: [180511221](https://github.com/dotnet/roslyn/pull/26055#discussion_r180511221) [](ancestors = 180511221)

AlekseyTs replied:
If the sorting is necessary, we can apply it in a separate targeted PR.

---
In reply to: [180514762](https://github.com/dotnet/roslyn/pull/26055#discussion_r180514762) [](ancestors = 180514762,180511221)

333fred replied:
Given that this was a generated change, I'm going to leave it as is. I did not manually change this file at all, this is purely codefix.

---
In reply to: [180522270](https://github.com/dotnet/roslyn/pull/26055#discussion_r180522270) [](ancestors = 180522270,180515588,180514762,180511221)

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -1264,11 +1267,12 @@ private IAddressOfOperation CreateBoundAddressOfOperatorOperation(BoundAddressOf
 
         private IInstanceReferenceOperation CreateBoundImplicitReceiverOperation(BoundImplicitReceiver boundImplicitReceiver)
         {
+            InstanceReferenceKind referenceKind = InstanceReferenceKind.ContainingTypeInstance;
```
>ContainingTypeInstance [](start = 72, length = 22)

Is this the right value? Should this be ```InitializedInstance``` instead?

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -759,19 +762,21 @@ Namespace Microsoft.CodeAnalysis.Operations
         End Function
 
         Private Function CreateBoundWithLValueExpressionPlaceholder(boundWithLValueExpressionPlaceholder As BoundWithLValueExpressionPlaceholder) As IInstanceReferenceOperation
+            Dim referenceKind As InstanceReferenceKind = InstanceReferenceKind.InitializedInstance
```
>InitializedInstance [](start = 79, length = 19)

I think this could be an implicit receiver in an With statement. We probably should have a special kind for this case and use the appropriate one here.

333fred replied:
Addressing this by renaming `InitializedInstance` to `ImplicitReceiver`. If we later get a request to be able to tell whether the instance is actually being initialized or just used as part of a With statement, then we can revisit or add some other boolean property.

---
In reply to: [180531576](https://github.com/dotnet/roslyn/pull/26055#discussion_r180531576) [](ancestors = 180531576)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -759,19 +762,21 @@ Namespace Microsoft.CodeAnalysis.Operations
         End Function
 
         Private Function CreateBoundWithRValueExpressionPlaceholder(boundWithRValueExpressionPlaceholder As BoundWithRValueExpressionPlaceholder) As IInstanceReferenceOperation
+            Dim referenceKind As InstanceReferenceKind = InstanceReferenceKind.InitializedInstance
```
>InitializedInstance [](start = 79, length = 19)

I think this could be an implicit receiver in an With statement. We probably should have a special kind for this case and use the appropriate one here.

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IWithStatement.vb:
```diff
@@ -38,7 +38,7 @@ IWithOperation (OperationKind.None, Type: null) (Syntax: 'With c'BIND ... End Wi
               IFieldReferenceOperation: C.I As System.Int32 (OperationKind.FieldReference, Type: System.Int32) (Syntax: '.I')
                 Instance Receiver: 
-                  IInstanceReferenceOperation (OperationKind.InstanceReference, Type: C, IsImplicit) (Syntax: 'c')
+                  IInstanceReferenceOperation (ReferenceKind: InitializedInstance) (OperationKind.InstanceReference, Type: C, IsImplicit) (Syntax: 'c')
```
>InitializedInstance [](start = 62, length = 19)

There is no initialized instance in this scenario

AlekseyTs commented on src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs:
```diff
@@ -872,6 +872,7 @@ public override void VisitParameterReference(IParameterReferenceOperation operat
         public override void VisitInstanceReference(IInstanceReferenceOperation operation)
```
>VisitInstanceReference [](start = 29, length = 22)

Consider making an adjustment to TestOperationVisitor as well. At least we should verify that the new property doesn't throw.

AlekseyTs commented:
Done with review pass (iteration 1). Tests are not looked at.

333fred commented:
This is object/collection initializers, or With statements. See the IOperationTests_IObjectCreationExpression.cs tests for a C# example.

---
In reply to: [380186314](https://github.com/dotnet/roslyn/pull/26055#issuecomment-380186314) [](ancestors = 380186314)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [380211901](https://github.com/dotnet/roslyn/pull/26055#issuecomment-380211901) [](ancestors = 380211901)

333fred commented:
Test failure is https://github.com/dotnet/roslyn/issues/12106

AlekseyTs commented on src/Compilers/Core/Portable/Operations/InstanceReferenceKind.cs:
```diff
@@ -0,0 +1,20 @@
+        /// </summary>
+        ContainingTypeInstance,
+        /// <summary>
+        /// Reference to the object being initialized in a C# object or collection initializer, or to the object being referred to in a VB With statement.
```
>in a C# object or collection initializer [](start = 54, length = 40)

VB also has these.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/IInstanceReferenceExpression.cs:
```diff
@@ -8,6 +8,8 @@ namespace Microsoft.CodeAnalysis.Operations
     /// Current usage:
     ///  (1) C# this or base expression.
     ///  (2) VB Me, MyClass, or MyBase expression.
+    ///  (3) C# object or collection initializers.
```
>C# object or collection initializers. [](start = 13, length = 37)

Why C# only? VB also has these.

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -1264,11 +1267,12 @@ private IAddressOfOperation CreateBoundAddressOfOperatorOperation(BoundAddressOf
 
         private IInstanceReferenceOperation CreateBoundImplicitReceiverOperation(BoundImplicitReceiver boundImplicitReceiver)
         {
+            InstanceReferenceKind referenceKind = InstanceReferenceKind.ImplicitReceiver;
```
>InstanceReferenceKind.ImplicitReceiver [](start = 50, length = 38)

Could you please point me to IOperation tests that go through this code path?

333fred replied:
Some of the tests I found:
* ObjectCreationWithArrayInitializer
* ObjectCreationWithInvalidMemberInitializer
* ObjectCreationWithInvalidInitializer
* DeconstructDelegate2
* ObjectInitializerTest_IncompleteComplexElementInitializerExpression
* CS0246_ERR_SingleTypeNameNotFound
* CS0117ERR_NoSuchMember
* AddPropertyUsedLikeMethod

---
In reply to: [180823954](https://github.com/dotnet/roslyn/pull/26055#discussion_r180823954) [](ancestors = 180823954)

333fred commented:
Ping @dotnet/roslyn-compiler @mavasani for reviews.

333fred commented:
Ping @dotnet/roslyn-compiler @mavasani for reviews.

mavasani commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IForLoopStatement.cs:
```diff
@@ -1686,14 +1686,14 @@ public class F
                                   IFieldReferenceOperation: System.Int32 F.i (OperationKind.FieldReference, Type: System.Int32) (Syntax: 'i')
                                     Instance Receiver: 
-                                      IInstanceReferenceOperation (OperationKind.InstanceReference, Type: F, IsImplicit) (Syntax: 'i')
+                                      IInstanceReferenceOperation (ReferenceKind: ImplicitReceiver) (OperationKind.InstanceReference, Type: F, IsImplicit) (Syntax: 'i')
```
Thanks, this is very useful information. Do we have an issue to track exposing a helper extension method that maps an `IInstanceReferenceOperation` with an `ImplicitReceiver` to the receiver operation? That will be a useful helper for dataflow analysis visitor, where when you encounter an `IInstanceReferenceOperation`, you normally want to return the tracked state/data for the receiver being referenced.

333fred replied:
Not currently, can you file one?

mavasani replied:
Filed https://github.com/dotnet/roslyn/issues/26152

333fred commented:
Ping @dotnet/roslyn-compiler for another review please

333fred commented:
@dotnet-bot retest windows_release_unit64_prtest please

AlekseyTs commented:
@cston, @dotnet/roslyn-compiler Please review. We need a second review for this PR, which **11 days old**.

333fred commented:
Test failures are https://github.com/dotnet/roslyn/issues/26037, the fix for which isn't in this branch yet.
